nidirect_cold_weather_payments/src/Form/ColdWeatherPaymentCheckerForm.php

```diff
@@ -184,40 +200,56 @@ public function cwpCheck(array $form, FormStateInterface $form_state) {
    */
   public function validateForm(array &$form, FormStateInterface $form_state) {
     parent::validateForm($form, $form_state);
+    // Validation is handled by ajax callback - or if no JS, by submitForm().
+  }
 
-    // Adding this validation to take care of older browsers.
-    $postcode = $form_state->getValue('postcode');
-
-    if (!empty($postcode)) {
-      $postcode_district = $this->cwpGetPostcodeDistrict($postcode);
-
-      if (!is_numeric($postcode_district) || $postcode_district < 1 || $postcode_district > 99) {
-        $form_state->setErrorByName('postcode', $this->t('Postcode must be a valid Northern Ireland postcode.'));
-      }
+  /**
+   * Validates that a postcode is a valid NI postcode or outward code (first half of the postcode).
+   */
+  protected function validatePostcode(array &$form, FormStateInterface $form_state) {
+    if (preg_match('/^BT[0-9]{1,2}( ?[0-9][A-Z]{2})?$/i', $form_state->getValue('postcode'))) {
+      return TRUE;
     }
+    return FALSE;
   }
 
   /**
    * {@inheritdoc}
    */
   public function submitForm(array &$form, FormStateInterface $form_state) {
-    $postcode = $form_state->getValue('postcode');
-    $postcode_district = $this->cwpGetPostcodeDistrict($postcode);
-    $data = $this->cwpLookup($postcode_district);
 
-    // Check we have data back from the API.
-    if (is_null($data)) {
-      $output = $this->t('Sorry, we were unable to process this request.');
+    // Since validation is handled by ajax callback, in the unlikely event that JS is disabled,
+    // validation will not have been performed. So do some basic validation here before processing the form submission.
+    $error = [
+      '#prefix' => '<p class="info-notice info-notice--error">',
+      '#markup' => $this->t('An error has occurred.'),
+      '#suffix' => '</p>',
+    ];
+
+    if (!$this->validatePostcode($form, $form_state)) {
+      $error['#markup'] = $this->t('Postcode must be a valid Northern Ireland postcode');
+      $output = $this->renderer->render($error);
     }
```

Comment:
This bit of validation is added here for rare cases where JS is disabled and validation didn't happen in the ajax callback ::cwpCheck()